kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/JvmDescriptorUtil.kt

```diff
@@ -0,0 +1,60 @@
+package com.squareup.moshi.kotlin.reflect
+
+import kotlinx.metadata.jvm.JvmFieldSignature
+import kotlinx.metadata.jvm.JvmMethodSignature
+import java.lang.reflect.Constructor
+import java.lang.reflect.Field
+import java.lang.reflect.Method
+
+internal val Class<*>.descriptor: String
+  get() {
+    return when {
+      isPrimitive -> when (kotlin) {
+        Byte::class -> "B"
+        Char::class -> "C"
+        Double::class -> "D"
+        Float::class -> "F"
+        Int::class -> "I"
+        Long::class -> "J"
+        Short::class -> "S"
+        Boolean::class -> "Z"
+        Void::class -> "V"
+        else -> throw RuntimeException("Unrecognized primitive $this")
+      }
+      isArray -> name.replace('.', '/')
+      else -> "L$name;".replace('.', '/')
+    }
+  }
+
+internal val Method.descriptor: String
+  get() = parameterTypes.joinToString(separator = "", prefix = "(", postfix = ")${returnType.descriptor}") { it.descriptor }
+
+/**
+ * Returns the JVM signature in the form "$Name$MethodDescriptor", for example: `equals(Ljava/lang/Object;)Z`.
+ *
+ * Useful for comparing with [JvmMethodSignature].
+ *
+ * For reference, see the [JVM specification, section 4.3](http://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3).
+ */
+internal val Method.jvmMethodSignature: String get() = "$name$descriptor"
+
+internal val Constructor<*>.descriptor: String
+  get() = parameterTypes.joinToString(separator = "", prefix = "(", postfix = ")V") { it.descriptor }
+
+/**
+ * Returns the JVM signature in the form "<init>$MethodDescriptor", for example: `"<init>(Ljava/lang/Object;)V")`.
+ *
+ * Useful for comparing with [JvmMethodSignature].
+ *
+ * For reference, see the [JVM specification, section 4.3](http://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3).
+ */
+internal val Constructor<*>.jvmMethodSignature: String get() = "<init>$descriptor"
+
+/**
+ * Returns the JVM signature in the form "$Name:$FieldDescriptor", for example: `"value:Ljava/lang/String;"`.
+ *
+ * Useful for comparing with [JvmFieldSignature].
+ *
+ * For reference, see the [JVM specification, section 4.3](http://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3).
+ */
+internal val Field.jvmFieldSignature: String get() = "$name:${type.descriptor}"
```

Comment:
Is this a real thing? I’ve never heard of a field signature before. (method signatures yes!)

Reply:
Yup. From the linked spec

> A descriptor is a string representing the type of a field or method. Descriptors are represented in the class file format using modified UTF-8 strings (§4.4.7) and thus may be drawn, where not further constrained, from the entire Unicode codespace.
>
> A signature is a string representing the generic type of a field or method, or generic type information for a class declaration.

Reply:
Resolving as it's been removed anyway